File: reflex_ui/reflex_ui/data/repository.py
from __future__ import annotations

import uuid
from collections.abc import Mapping
from typing import Any

from psycopg import IntegrityError
from psycopg.rows import dict_row
from psycopg.types.json import Json

from reflex_ui.data import auth_crypto


def _normalize_email(email: str) -> str:
    return email.strip().lower()


async def insert_user(
    pool, email: str, password: str, display_name: str | None
) -> uuid.UUID:
    email_n = _normalize_email(email)
    ph = auth_crypto.hash_password(password)
    async with pool.connection() as conn:
        async with conn.transaction():
            async with conn.cursor() as cur:
                await cur.execute(
                    """
                    INSERT INTO users (email, display_name, password_hash)
                    VALUES (%s, %s, %s)
                    RETURNING id
                    """,
                    (email_n, display_name, ph),
                )
                row = await cur.fetchone()
    if not row:
        raise RuntimeError("User insert did not return id")
    return row[0]


async def insert_user_safe(
    pool, email: str, password: str, display_name: str | None
) -> uuid.UUID | None:
    try:
        return await insert_user(pool, email, password, display_name)
    except IntegrityError:
        return None


async def fetch_user_by_email(pool, email: str) -> dict[str, Any] | None:
    email_n = _normalize_email(email)
    async with pool.connection() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, email, display_name, password_hash, created_at, updated_at
                FROM users
                WHERE email = %s
                """,
                (email_n,),
            )
            row = await cur.fetchone()
            return dict(row) if row else None


async def fetch_user_by_id(pool, user_id: uuid.UUID) -> dict[str, Any] | None:
    async with pool.connection() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, email, display_name, password_hash, created_at, updated_at
                FROM users
                WHERE id = %s
                """,
                (user_id,),
            )
            row = await cur.fetchone()
            return dict(row) if row else None


async def authenticate_user(pool, email: str, password: str) -> uuid.UUID | None:
    row = await fetch_user_by_email(pool, email)
    if not row:
        return None
    if not auth_crypto.verify_password(password, row["password_hash"]):
        return None
    return row["id"]


async def count_chat_sessions(pool, user_id: uuid.UUID) -> int:
    async with pool.connection() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                SELECT COUNT(*)::int
                FROM chat_sessions
                WHERE user_id = %s AND archived_at IS NULL
                """,
                (user_id,),
            )
            row = await cur.fetchone()
            return int(row[0]) if row else 0


async def list_chat_sessions(
    pool, user_id: uuid.UUID, *, limit: int, offset: int
) -> list[dict[str, Any]]:
    async with pool.connection() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, title, created_at, updated_at
                FROM chat_sessions
                WHERE user_id = %s AND archived_at IS NULL
                ORDER BY updated_at DESC
                LIMIT %s OFFSET %s
                """,
                (user_id, limit, offset),
            )
            rows = await cur.fetchall()
            return [dict(r) for r in rows]


async def list_all_chat_sessions(
    pool, user_id: uuid.UUID
) -> list[dict[str, Any]]:
    async with pool.connection() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, title, created_at, updated_at
                FROM chat_sessions
                WHERE user_id = %s AND archived_at IS NULL
                ORDER BY updated_at DESC
                """,
                (user_id,),
            )
            rows = await cur.fetchall()
            return [dict(r) for r in rows]


async def create_chat_session(
    pool, user_id: uuid.UUID, title: str = "New Chat"
) -> uuid.UUID:
    async with pool.connection() as conn:
        async with conn.transaction():
            async with conn.cursor() as cur:
                await cur.execute(
                    """
                    INSERT INTO chat_sessions (user_id, title)
                    VALUES (%s, %s)
                    RETURNING id
                    """,
                    (user_id, title),
                )
                row = await cur.fetchone()
    if not row:
        raise RuntimeError("Session insert did not return id")
    return row[0]


async def update_chat_session_title(
    pool, user_id: uuid.UUID, session_id: uuid.UUID, title: str
) -> bool:
    t = (title or "").strip()
    if not t:
        return False
    async with pool.connection() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                UPDATE chat_sessions
                SET title = %s
                WHERE id = %s AND user_id = %s AND archived_at IS NULL
                """,
                (t, session_id, user_id),
            )
            return cur.rowcount > 0


async def assert_session_owned(
    pool, user_id: uuid.UUID, session_id: uuid.UUID
) -> bool:
    async with pool.connection() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                SELECT 1
                FROM chat_sessions
                WHERE id = %s AND user_id = %s AND archived_at IS NULL
                """,
                (session_id, user_id),
            )
            return await cur.fetchone() is not None


async def delete_chat_session(pool, user_id: uuid.UUID, session_id: uuid.UUID) -> bool:
    async with pool.connection() as conn:
        async with conn.transaction():
            async with conn.cursor() as cur:
                await cur.execute(
                    """
                    DELETE FROM chat_sessions
                    WHERE id = %s AND user_id = %s
                    """,
                    (session_id, user_id),
                )
                deleted = cur.rowcount
    return deleted > 0


async def list_messages(
    pool, user_id: uuid.UUID, session_id: uuid.UUID
) -> list[dict[str, Any]]:
    if not await assert_session_owned(pool, user_id, session_id):
        return []
    async with pool.connection() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, session_id, role, content, model, token_usage_json,
                       sequence_no, created_at
                FROM chat_messages
                WHERE session_id = %s
                ORDER BY sequence_no ASC, created_at ASC
                """,
                (session_id,),
            )
            rows = await cur.fetchall()
            return [dict(r) for r in rows]


async def insert_message(
    pool,
    user_id: uuid.UUID,
    session_id: uuid.UUID,
    *,
    role: str,
    content: str,
    model: str | None = None,
    token_usage: Mapping[str, Any] | None = None,
) -> uuid.UUID | None:
    async with pool.connection() as conn:
        async with conn.transaction():
            async with conn.cursor() as cur:
                await cur.execute(
                    """
                    SELECT 1
                    FROM chat_sessions
                    WHERE id = %s AND user_id = %s AND archived_at IS NULL
                    FOR UPDATE
                    """,
                    (session_id, user_id),
                )
                if await cur.fetchone() is None:
                    return None
                usage = Json(dict(token_usage)) if token_usage is not None else None
                await cur.execute(
                    """
                    INSERT INTO chat_messages (
                        session_id, role, content, model, token_usage_json, sequence_no
                    )
                    VALUES (
                        %s, %s, %s, %s, %s,
                        (SELECT COALESCE(MAX(sequence_no), 0) + 1
                         FROM chat_messages WHERE session_id = %s)
                    )
                    RETURNING id
                    """,
                    (session_id, role, content, model, usage, session_id),
                )
                row = await cur.fetchone()
                await cur.execute(
                    """
                    UPDATE chat_sessions
                    SET updated_at = NOW()
                    WHERE id = %s
                    """,
                    (session_id,),
                )
    return row[0] if row else None

File: scienta_ui/src/lib/repository.ts
import { randomUUID } from "node:crypto";
import type { PoolClient, QueryResultRow } from "pg";
import { hashPassword, verifyPassword } from "@/lib/auth";
import { getPool } from "@/lib/db";
import { formatMessageTimePtBr, formatSessionSidebarDate } from "@/lib/formatting";
export type SessionDto = {
  id: string;
  title: string;
  title_short: string;
  updated_at: string;
  created_at: string;
  updated_display: string;
};

export type ThreadMessageDto = {
  id: string;
  role: "user" | "assistant";
  content: string;
  created_at: string;
  time_display: string;
};


function normalizeEmail(email: string): string {
  return email.trim().toLowerCase();
}

type DbSession = {
  id: string;
  title: string;
  created_at: Date | string;
  updated_at: Date | string;
};

type DbMessage = {
  id: string;
  role: string;
  content: string;
  model: string | null;
  sequence_no: number;
  created_at: Date | string;
  token_usage_json: unknown;
};

function mapSessionRow(row: DbSession): SessionDto {
  const title = row.title || "New Chat";
  const titleShort = title.length <= 30 ? title : `${title.slice(0, 27)}...`;
  const updatedIso = new Date(row.updated_at).toISOString();
  const createdIso = new Date(row.created_at).toISOString();
  return {
    id: row.id,
    title,
    title_short: titleShort,
    updated_at: updatedIso,
    created_at: createdIso,
    updated_display: formatSessionSidebarDate(updatedIso),
  };
}

function mapThreadMessageRow(row: DbMessage): ThreadMessageDto {
  const createdAt = new Date(row.created_at).toISOString();
  return {
    id: row.id,
    role: (row.role === "assistant" ? "assistant" : "user") as "user" | "assistant",
    content: row.content,
    created_at: createdAt,
    time_display: formatMessageTimePtBr(createdAt),
  };
}

function isUniqueViolation(error: unknown): boolean {
  if (!error || typeof error !== "object") {
    return false;
  }
  const maybeCode = (error as { code?: string }).code;
  return maybeCode === "23505";
}

export async function insertUserSafe(
  email: string,
  password: string,
  displayName: string | null
): Promise<string | null> {
  const pool = await getPool();
  try {
    const result = await pool.query<{ id: string }>(
      `INSERT INTO users (email, display_name, password_hash)
       VALUES ($1, $2, $3)
       RETURNING id`,
      [normalizeEmail(email), displayName, hashPassword(password)]
    );
    return result.rows[0]?.id ?? null;
  } catch (error) {
    if (isUniqueViolation(error)) {
      return null;
    }
    throw error;
  }
}

export async function fetchUserByEmail(email: string): Promise<QueryResultRow | null> {
  const pool = await getPool();
  const result = await pool.query(
    `SELECT id, email, display_name, password_hash, created_at, updated_at
     FROM users
     WHERE email = $1`,
    [normalizeEmail(email)]
  );
  return result.rows[0] ?? null;
}

export async function fetchUserById(userId: string): Promise<QueryResultRow | null> {
  const pool = await getPool();
  const result = await pool.query(
    `SELECT id, email, display_name, password_hash, created_at, updated_at
     FROM users
     WHERE id = $1`,
    [userId]
  );
  return result.rows[0] ?? null;
}

export async function authenticateUser(email: string, password: string): Promise<string | null> {
  const row = await fetchUserByEmail(email);
  if (!row?.password_hash || typeof row.password_hash !== "string") {
    return null;
  }
  if (!verifyPassword(password, row.password_hash)) {
    return null;
  }
  return String(row.id);
}

export async function listAllChatSessions(userId: string): Promise<SessionDto[]> {
  const pool = await getPool();
  const result = await pool.query<DbSession>(
    `SELECT id, title, created_at, updated_at
     FROM chat_sessions
     WHERE user_id = $1 AND archived_at IS NULL
     ORDER BY updated_at DESC`,
    [userId]
  );
  return result.rows.map(mapSessionRow);
}

export async function assertSessionOwned(userId: string, sessionId: string): Promise<boolean> {
  const pool = await getPool();
  const result = await pool.query(
    `SELECT 1
     FROM chat_sessions
     WHERE id = $1 AND user_id = $2 AND archived_at IS NULL`,
    [sessionId, userId]
  );
  return Boolean(result.rows[0]);
}

export async function createChatSession(userId: string, title = "New Chat"): Promise<string> {
  const pool = await getPool();
  const result = await pool.query<{ id: string }>(
    `INSERT INTO chat_sessions (user_id, title)
     VALUES ($1, $2)
     RETURNING id`,
    [userId, title]
  );
  return result.rows[0]?.id ?? randomUUID();
}

export async function updateChatSessionTitle(
  userId: string,
  sessionId: string,
  title: string
): Promise<boolean> {
  const t = title.trim();
  if (!t) {
    return false;
  }
  const pool = await getPool();
  const result = await pool.query(
    `UPDATE chat_sessions
     SET title = $1
     WHERE id = $2 AND user_id = $3 AND archived_at IS NULL`,
    [t, sessionId, userId]
  );
  return (result.rowCount ?? 0) > 0;
}

export async function deleteChatSession(userId: string, sessionId: string): Promise<boolean> {
  const pool = await getPool();
  const result = await pool.query(
    `DELETE FROM chat_sessions
     WHERE id = $1 AND user_id = $2`,
    [sessionId, userId]
  );
  return (result.rowCount ?? 0) > 0;
}

export async function listMessages(userId: string, sessionId: string): Promise<ThreadMessageDto[]> {
  if (!(await assertSessionOwned(userId, sessionId))) {
    return [];
  }
  const pool = await getPool();
  const result = await pool.query<DbMessage>(
    `SELECT id, session_id, role, content, model, token_usage_json, sequence_no, created_at
     FROM chat_messages
     WHERE session_id = $1
     ORDER BY sequence_no ASC, created_at ASC`,
    [sessionId]
  );
  return result.rows.map(mapThreadMessageRow);
}

async function withSessionLock<T>(
  userId: string,
  sessionId: string,
  fn: (client: PoolClient) => Promise<T>
): Promise<T | null> {
  const pool = await getPool();
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    const ownership = await client.query(
      `SELECT 1
       FROM chat_sessions
       WHERE id = $1 AND user_id = $2 AND archived_at IS NULL
       FOR UPDATE`,
      [sessionId, userId]
    );
    if (!ownership.rowCount) {
      await client.query("ROLLBACK");
      return null;
    }
    const result = await fn(client);
    await client.query("COMMIT");
    return result;
  } catch (error) {
    await client.query("ROLLBACK");
    throw error;
  } finally {
    client.release();
  }
}

export async function insertMessage(args: {
  userId: string;
  sessionId: string;
  role: string;
  content: string;
  model?: string | null;
  tokenUsage?: Record<string, unknown> | null;
}): Promise<string | null> {
  const insertedId = await withSessionLock(args.userId, args.sessionId, async (client) => {
    const result = await client.query<{ id: string }>(
      `INSERT INTO chat_messages (session_id, role, content, model, token_usage_json, sequence_no)
       VALUES (
         $1, $2, $3, $4, $5,
         (SELECT COALESCE(MAX(sequence_no), 0) + 1 FROM chat_messages WHERE session_id = $1)
       )
       RETURNING id`,
      [
        args.sessionId,
        args.role,
        args.content,
        args.model ?? null,
        args.tokenUsage ?? null,
      ]
    );
    await client.query(
      `UPDATE chat_sessions
       SET updated_at = NOW()
       WHERE id = $1`,
      [args.sessionId]
    );
    return result.rows[0]?.id ?? null;
  });
  return insertedId;
}
